Errors updating to 0.9.1 - key not found and method overloads mismatch

Not entirely sure this is a mima issue, but, the [PR upgrading Akka to MiMa 0.9.1](https://travis-ci.com/github/akka/akka/jobs/508193351) failed with these two errors: 

```
[error] java.util.NoSuchElementException: key not found: <noclass>$inflating$
[error] 	at scala.collection.MapLike.default(MapLike.scala:235)
[error] 	at scala.collection.MapLike.default$(MapLike.scala:234)
[error] 	at scala.collection.AbstractMap.default(Map.scala:65)
[error] 	at scala.collection.MapLike.apply(MapLike.scala:144)
[error] 	at scala.collection.MapLike.apply$(MapLike.scala:143)
[error] 	at scala.collection.AbstractMap.apply(Map.scala:65)
[error] 	at com.typesafe.tools.mima.core.MimaUnpickler$.withOwner$1(MimaUnpickler.scala:48)
[error] 	at com.typesafe.tools.mima.core.MimaUnpickler$.symbolToClass$1(MimaUnpickler.scala:54)
[error] 	at com.typesafe.tools.mima.core.MimaUnpickler$.$anonfun$unpickleClass$16(MimaUnpickler.scala:84)
[error] 	at com.typesafe.tools.mima.core.MimaUnpickler$.$anonfun$unpickleClass$16$adapted(MimaUnpickler.scala:83)
[error] 	at scala.collection.Iterator.foreach(Iterator.scala:943)
[error] 	at scala.collection.Iterator.foreach$(Iterator.scala:943)
[error] 	at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
[error] 	at com.typesafe.tools.mima.core.MimaUnpickler$.unpickleClass(MimaUnpickler.scala:83)
[error] 	at com.typesafe.tools.mima.core.ClassfileParser.parsePickle(ClassfileParser.scala:147)
[error] 	at com.typesafe.tools.mima.core.ClassfileParser.parseClassAttributes(ClassfileParser.scala:64)
[error] 	at com.typesafe.tools.mima.core.ClassfileParser.com$typesafe$tools$mima$core$ClassfileParser$$parseClass(ClassfileParser.scala:19)
[error] 	at com.typesafe.tools.mima.core.ClassfileParser$.parseInPlace(ClassfileParser.scala:168)
[error] 	at com.typesafe.tools.mima.core.ConcreteClassInfo.afterLoading(ClassInfo.scala:44)
[error] 	at com.typesafe.tools.mima.core.ClassInfo.flags(ClassInfo.scala:76)
[error] 	at com.typesafe.tools.mima.core.InfoLike.isPublic(InfoLike.scala:14)
[error] 	at com.typesafe.tools.mima.core.PackageInfo.isAccessible$1(PackageInfo.scala:81)
[error] 	at com.typesafe.tools.mima.core.PackageInfo.$anonfun$accessibleClasses$1(PackageInfo.scala:75)
[error] 	at com.typesafe.tools.mima.core.PackageInfo.$anonfun$accessibleClasses$1$adapted(PackageInfo.scala:75)
```

and 

```
java.lang.AssertionError: assertion failed: method overloads mismatch; bytecode=List(def mergeDisjointKeys: (Lscala.collection.immutable.Set;Lscala.collection.immutable.Map;Lakka.persistence.typed.internal.VersionVector;Lscala.collection.immutable.Map;)Lscala.collection.immutable.Map;) pickle=Stream(SymbolInfo(VALsym, mergeDisjointKeys, owner=1, isScopedPrivate=true), SymbolInfo(VALsym, mergeDisjointKeys, owner=1, isScopedPrivate=false)) for object akka.persistence.typed.crdt.ORSet
[error] 	at scala.Predef$.assert(Predef.scala:223)
[error] 	at com.typesafe.tools.mima.core.MimaUnpickler$.$anonfun$unpickleClass$8(MimaUnpickler.scala:66)
[error] 	at com.typesafe.tools.mima.core.MimaUnpickler$.$anonfun$unpickleClass$8$adapted(MimaUnpickler.scala:63)
[error] 	at scala.collection.immutable.HashMap$HashMap1.foreach(HashMap.scala:399)
[error] 	at scala.collection.immutable.HashMap$HashTrieMap.foreach(HashMap.scala:725)
[error] 	at com.typesafe.tools.mima.core.MimaUnpickler$.doMethods$1(MimaUnpickler.scala:63)
[error] 	at com.typesafe.tools.mima.core.MimaUnpickler$.$anonfun$unpickleClass$18(MimaUnpickler.scala:91)
[error] 	at com.typesafe.tools.mima.core.MimaUnpickler$.$anonfun$unpickleClass$18$adapted(MimaUnpickler.scala:86)
```

--- a/core/src/main/scala/com/typesafe/tools/mima/core/MimaUnpickler.scala
+++ b/core/src/main/scala/com/typesafe/tools/mima/core/MimaUnpickler.scala
@@ -31,30 +31,29 @@ object MimaUnpickler {
       SymbolInfo(tag, name, owner, flags, isScopedPrivate)
     }
 
-    def symbolToClass(symbolInfo: SymbolInfo) = {
+    def symbolToClass(symbolInfo: SymbolInfo): ClassInfo = {
       if (symbolInfo.name == REFINE_CLASS_NAME) {
         // eg: CLASSsym 4: 89(<refinement>) 0 0[] 87
         // Nsc's UnPickler also excludes these with "isRefinementSymbolEntry"
         NoClass
-      } else if (symbolInfo.name == "<local child>") {
+      } else if (symbolInfo.name == LOCAL_CHILD) {
         // Predef$$less$colon$less$<local child>
         NoClass
       } else {
-        val own = classes(symbolInfo.owner)
-        def withOwner(cls: ClassInfo) = {
-          val nme1 = cls.bytecodeName
-          val nme2 = symbolInfo.name
-          val nme3 = if (nme2.startsWith(nme1)) nme2.substring(nme2.lastIndexOf('$') + 1) else nme2
-          val conc = if (nme1.endsWith("$")) "" else "$"
-          val suff = if (symbolInfo.isModuleOrModuleClass) "$" else ""
-          val name = nme1 + conc + nme3 + suff
-          clazz.owner.classes(name)
-        }
         val fallback = if (symbolInfo.isModuleOrModuleClass) clazz.moduleClass else clazz
-        own match {
-          case null if symbolInfo.owner == 0 => withOwner(fallback)
+        def lookup(cls: ClassInfo) = {
+          val clsName   = cls.bytecodeName
+          val separator = if (clsName.endsWith("$")) "" else "$"
+          val name0     = symbolInfo.name
+          val name      = if (name0.startsWith(clsName)) name0.substring(name0.lastIndexOf('$') + 1) else name0
+          val suffix    = if (symbolInfo.isModuleOrModuleClass) "$" else ""
+          val newName   = clsName + separator + name + suffix
+          clazz.owner.classes.getOrElse(newName, NoClass)
+        }
+        classes(symbolInfo.owner) match {
+          case null if symbolInfo.owner == 0 => lookup(fallback)
           case null                          => fallback
-          case cls                           => withOwner(cls)
+          case cls                           => lookup(cls)
         }
       }
     }
@@ -62,27 +61,30 @@ object MimaUnpickler {
     def doMethods(clazz: ClassInfo, methods: List[SymbolInfo]) = {
       methods.iterator
         .filter(!_.isParam)
-        .filter(_.name != "<init>") // TODO support package private constructors
+        .filter(_.name != CONSTRUCTOR) // TODO support package private constructors
         .toSeq.groupBy(_.name).foreach { case (name, pickleMethods) =>
-          val bytecodeMethods = clazz.methods.get(name).filter(!_.isBridge).toList
-          // #630 one way this happens with mixins:
-          //    trait Foo { def bar(x: Int): Int = x }
-          //    class Bar extends Foo { private[foo] def bar: String = "" }
-          // during pickling Bar only contains the package private bar()String
-          // but later in the backend the classfile gets a copy of bar(Int)Int
-          // so the "bar" method in the pickle doesn't know which bytecode method it's about
-          // the proper way to fix this involves unpickling the types in the pickle,
-          // then implementing the rules of erasure, so that you can then match the pickle
-          // types with the erased types.  Meanwhile we'll just ignore them, worst case users
-          // need to add a filter like they have for years.
-          if (pickleMethods.size == bytecodeMethods.size && pickleMethods.exists(_.isScopedPrivate)) {
-            bytecodeMethods.zip(pickleMethods).foreach { case (bytecodeMeth, pickleMeth) =>
-              bytecodeMeth.scopedPrivate = pickleMeth.isScopedPrivate
-            }
-          }
+          doMethodOverloads(clazz, name, pickleMethods)
       }
     }
 
+    def doMethodOverloads(clazz: ClassInfo, name: String, pickleMethods: Seq[SymbolInfo]) = {
+      val bytecodeMethods = clazz.methods.get(name).filter(!_.isBridge).toList
+      // #630 one way this happens with mixins:
+      //    trait Foo { def bar(x: Int): Int = x }
+      //    class Bar extends Foo { private[foo] def bar: String = "" }
+      // during pickling Bar only contains the package private bar()String
+      // but later in the backend the classfile gets a copy of bar(Int)Int
+      // so the "bar" method in the pickle doesn't know which bytecode method it's about
+      // the proper way to fix this involves unpickling the types in the pickle,
+      // then implementing the rules of erasure, so that you can then match the pickle
+      // types with the erased types.  Meanwhile we'll just ignore them, worst case users
+      // need to add a filter like they have for years.
+      if (pickleMethods.size == bytecodeMethods.size && pickleMethods.exists(_.isScopedPrivate)) {
+        bytecodeMethods.zip(pickleMethods).foreach { case (bytecodeMeth, pickleMeth) =>
+          bytecodeMeth.scopedPrivate = pickleMeth.isScopedPrivate
+        }
+      }
+    }
 
     for (num <- index.indices) {
       buf.atIndex(index(num)) {
@@ -157,5 +159,7 @@ object MimaUnpickler {
     final val PARAM      = 1L << 13
   }
 
+  final val CONSTRUCTOR       = "<init>"
+  final val LOCAL_CHILD       = "<local child>"
   final val REFINE_CLASS_NAME = "<refinement>"
 }

